File: calibration/nersc/run_generation.m
function run_generation()
%RUN_GENERATION  作业期间常驻的标定驱动 (NERSC 单节点 + parpool)。
%
%   一个 MATLAB 进程持有一个 parpool, 跨代复用 (不反复拉起 MATLAB)。
%   每代: 读 state.json 的 batchNN_cases -> parfeval 提交 nCase*3 条腿 ->
%         全部回收后逐 case 抽指标 -> 判成功 -> 调 CMA-ES 出下一代 -> 继续。
%
%   环境变量:
%     CALIB_REPO_DIR     仓库根 (默认 = 本文件上上层)
%     CALIB_WORKERS      pool worker 数 (默认 90 = 30 case x 3 dose)
%     CALIB_POPULATION   每代 case 数 (默认 30; 同时传给 CMA-ES)
%     CALIB_WALL_BUDGET  本作业可用秒数 (submit.sh 设为 walltime - 余量)
%     CALIB_GEN_RESERVE  开新一代所需的最少剩余秒数 (默认 5400)
%     CALIB_ENDPOINT_TOL 成功判据的端点容差 nm (默认 3)
%     CALIB_PYTHON       python3 可执行文件 (默认 python3)
%
%   墙钟耗尽: 腿自己按 WALL_BUDGET 存 checkpoint 退出, 本函数干净返回,
%   submit.sh 续投下一个作业, 未完成的腿从 checkpoint 续算 (幂等)。

here = fileparts(mfilename('fullpath'));
repo = getenv('CALIB_REPO_DIR');
if isempty(repo), repo = fileparts(fileparts(here)); end
addpath(repo);  addpath(here);

calib   = fullfile(repo, 'calibration');
statef  = fullfile(calib, 'state.json');
logdir  = fullfile(calib, 'logs');
if ~exist(logdir, 'dir'), mkdir(logdir); end

budget   = envnum('CALIB_WALL_BUDGET',  inf);
reserve  = envnum('CALIB_GEN_RESERVE',  5400);
workers  = envnum('CALIB_WORKERS',      90);
popsize  = envnum('CALIB_POPULATION',   30);
tol      = envnum('CALIB_ENDPOINT_TOL', 3);
python   = getenv('CALIB_PYTHON');
if isempty(python), python = 'python3'; end
gentool  = fullfile(here, 'gen_tool.py');

tStart = tic;
if isfinite(budget)
    say('[boot] repo=%s workers=%d pop=%d budget=%.0f s', repo, workers, popsize, budget);
else
    say('[boot] repo=%s workers=%d pop=%d 无代码侧时限 (跑到 SLURM 杀)', ...
        repo, workers, popsize);
end

% ---------- parpool: 整个作业只开一次 ----------
pool = gcp('nocreate');
if isempty(pool)
    cl = parcluster('Processes');
    store = fullfile(scratchdir(), 'matlab_jobstore', ...
                     sprintf('job%s_pid%d', jobid(), feature('getpid')));
    if ~exist(store, 'dir'), mkdir(store); end
    cl.JobStorageLocation = store;
    cl.NumWorkers = workers;
    pool = parpool(cl, workers, 'IdleTimeout', inf);
end
say('[pool] %d workers 就绪, 耗时 %.0f s', pool.NumWorkers, toc(tStart));

% ---------- 代循环 ----------
while true
    % 哨兵: 命中或人工停止 (另一个作业写的 DONE, 或你 touch 的 STOP)
    if exist(fullfile(calib,'DONE'),'file') || exist(fullfile(calib,'STOP'),'file')
        say('[sentinel] 见到 DONE/STOP, 退出');
        break
    end
    % 默认 budget=inf: 不自我限时, 跑到 SLURM 杀为止。
    % 只有显式设了 CALIB_WALL_BUDGET 才会提前收手。
    if isfinite(budget) && toc(tStart) > budget - reserve
        say('[wall] 剩余不足一代 (%.0f s), 干净退出', budget - toc(tStart));
        break
    end

    batch = next_batch(statef);

    % 续跑摘要: 打印 CMA 持久状态 + 全历史 fitness (作业重启后一眼看出从哪捡起)
    [~, hist] = sh(sprintf('%s %s history %d', python, q(gentool), batch), ...
                   repo, popsize, python);
    fprintf('%s', hist);

    % 提议本代 (幂等: pending_batch 命中则原样重发, 不推进 CMA)
    rc = sh(sprintf('%s %s propose %d', python, q(gentool), batch), ...
            repo, popsize, python);
    if rc ~= 0 && batch > 1
        % 常见死锁: 上一代有 case 既没指标也没被熔断 -> CMA 拒绝 tell。
        % 一次性惩罚这些 case 后重试, 避免续投链无限重复同一个失败。
        say('[propose] 失败, 尝试惩罚上一代缺指标的 case 后重试');
        sh(sprintf('%s %s penalize-missing %d', python, q(gentool), batch-1), ...
           repo, popsize, python);
        rc = sh(sprintf('%s %s propose %d', python, q(gentool), batch), ...
                repo, popsize, python);
    end
    if rc ~= 0
        error('run_generation:propose', 'CMA-ES 提议失败, batch %d', batch);
    end
    cases = read_cases(statef, batch);
    say('[gen %02d] %d cases x 3 dose = %d legs', batch, numel(cases), 3*numel(cases));

    % ---- 撒腿 ----
    doses = [0, 0.5, 3];
    if isfinite(budget)
        legBudget = max(600, budget - toc(tStart) - 300);   % 腿自留 5 min 存盘
    else
        legBudget = inf;                                   % 不限时
    end
    F = parallel.FevalFuture.empty(0,1);
    meta = {};
    for i = 1:numel(cases)
        tag  = cases(i).case_tag;
        mult = cases(i).mult(:)';
        for d = doses
            if leg_is_complete(repo,tag, d), continue; end
            F(end+1,1) = parfeval(pool, @run_leg_worker, 1, ...
                                  repo, tag, d, mult, legBudget); %#ok<AGROW>
            meta{end+1} = sprintf('%s/dose%g', tag, d);            %#ok<AGROW>
        end
    end
    nTotal = numel(F);
    if nTotal == 0
        say('[gen %02d] 所有腿已完成 (续投恢复), 直接进入抽指标', batch);
    end

    % ---- 回收 (worker 内部已 try/catch, fetchNext 不会抛) ----
    attemptsFile = fullfile(here, 'attempts.json');
    A = load_attempts(attemptsFile);
    maxAttempts = envnum('CALIB_MAX_ATTEMPTS', 3);
    nDone = 0;  tGen = tic;
    while nDone < nTotal
        [idx, status] = fetchNext(F, 120);
        if isempty(idx)
            say('[gen %02d] 进行中 %d/%d, 本代已用 %.0f min', ...
                batch, nDone, nTotal, toc(tGen)/60);
            continue
        end
        nDone = nDone + 1;
        if startsWith(status, 'error')
            k = keyof(meta{idx});
            A.(k) = getfielddef(A, k, 0) + 1;
            save_attempts(attemptsFile, A);
            say('[leg] %-24s %s  (第 %d 次失败)', meta{idx}, status, A.(k));
        else
            say('[leg] %-24s %-10s (%d/%d, 本代 %.0f min)', ...
                meta{idx}, status, nDone, nTotal, toc(tGen)/60);
        end
    end

    % ---- 熔断: 反复失败的 case 写惩罚指标, 让 CMA 判它不可行并前进 ----
    dead = {};
    for i = 1:numel(cases)
        tag = cases(i).case_tag;
        for d = doses
            k = keyof(sprintf('%s/dose%g', tag, d));
            if ~leg_is_complete(repo,tag, d) && getfielddef(A, k, 0) >= maxAttempts
                dead{end+1} = tag; %#ok<AGROW>
                break
            end
        end
    end
    dead = unique(dead);
    for i = 1:numel(dead)
        sh(sprintf('%s %s penalize %d --tag %s', python, q(gentool), batch, dead{i}), ...
           repo, popsize, python);
        say('[dead] %s 连续失败 >=%d 次, 已写惩罚指标', dead{i}, maxAttempts);
    end

    % ---- 完整性: 有腿没跑完且未熔断 = 墙钟耗尽, 交给续投 ----
    incomplete = 0;
    for i = 1:numel(cases)
        if any(strcmp(cases(i).case_tag, dead)), continue; end
        for d = doses
            if ~leg_is_complete(repo,cases(i).case_tag, d), incomplete = incomplete + 1; end
        end
    end
    if incomplete > 0
        say('[gen %02d] %d 条腿未完成 (墙钟), 存盘退出等续投', batch, incomplete);
        break
    end

    % ---- 抽指标 ----
    for i = 1:numel(cases)
        tag = cases(i).case_tag;
        if exist(fullfile(calib,'metrics',[tag '.csv']), 'file'), continue; end
        try
            extract_calibration_metrics(tag);
        catch err
            say('[metrics] %s 失败: %s', tag, err.message);
            sh(sprintf('%s %s penalize %d --tag %s', python, q(gentool), batch, tag), ...
               repo, popsize, python);
        end
    end

    % ---- 判成功 ----
    [rc, out] = sh(sprintf('%s %s success %d --tol %g', ...
                           python, q(gentool), batch, tol), repo, popsize, python);
    say('[gen %02d] %s', batch, strtrim(out));
    w = regexp(out, 'WINNERS=(\S+)', 'tokens', 'once');
    if rc == 0 && ~isempty(w) && ~isempty(strtrim(w{1}))
        fid = fopen(fullfile(calib, 'DONE'), 'w');
        fprintf(fid, '%s\n', strtrim(w{1}));  fclose(fid);
        say('[done] 命中: %s', strtrim(w{1}));
        break
    end
    sh(sprintf('%s %s report %d', python, q(gentool), batch), repo, popsize, python);
end

say('[exit] 总耗时 %.1f h', toc(tStart)/3600);
end

% ===================================================================
function b = next_batch(statef)
% 已有的最大 batch: 若其 cases 已全部抽完指标则 +1, 否则续跑它。
b = 1;
if ~exist(statef, 'file'), return; end
s = jsondecode(fileread(statef));
f = fieldnames(s);
nums = [];
for i = 1:numel(f)
    t = regexp(f{i}, '^batch(\d+)_cases$', 'tokens', 'once');
    if ~isempty(t), nums(end+1) = str2double(t{1}); end %#ok<AGROW>
end
if isempty(nums), return; end
b = max(nums);
calib = fileparts(statef);
cases = read_cases(statef, b);
for i = 1:numel(cases)
    if ~exist(fullfile(calib,'metrics',[cases(i).case_tag '.csv']), 'file')
        return                      % 本代还没做完, 续跑本代
    end
end
b = b + 1;                          % 本代已完结, 进下一代
end

function cases = read_cases(statef, batch)
s = jsondecode(fileread(statef));
key = sprintf('batch%02d_cases', batch);
if ~isfield(s, key)
    error('run_generation:nocases', 'state.json 缺 %s', key);
end
c = s.(key);
if iscell(c), c = [c{:}]; end
cases = c(:)';
end

% 完成判据统一走仓库根的 leg_is_complete.m (要求 _COMPLETE 标记), 此处不再本地实现

function [rc, out] = sh(cmd, repo, popsize, ~)
env = sprintf(['CALIB_REPO_DIR=%s CALIB_POPULATION=%d '], q(repo), popsize);
[rc, out] = system([env ' ' cmd]);
end

function A = load_attempts(f)
A = struct();
if exist(f, 'file')
    try, A = jsondecode(fileread(f)); catch, A = struct(); end
end
end

function save_attempts(f, A)
try
    fid = fopen([f '.tmp'], 'w');  fprintf(fid, '%s', jsonencode(A));  fclose(fid);
    movefile([f '.tmp'], f, 'f');
catch
end
end

function k = keyof(s)
k = regexprep(s, '[^A-Za-z0-9]', '_');   % 'b01_c01_cma/dose0.5' -> 合法字段名
end

function v = getfielddef(S, k, default)
if isfield(S, k), v = S.(k); else, v = default; end
end

function v = envnum(name, default)
v = str2double(getenv(name));
if isnan(v), v = default; end
end

function s = q(p), s = ['"' char(p) '"']; end

function d = scratchdir()
d = getenv('SCRATCH');
if isempty(d) || ~exist(d,'dir'), d = tempdir; end
end

function s = jobid()
s = getenv('SLURM_JOB_ID');
if isempty(s), s = 'local'; end
end

function say(fmt, varargin)
fprintf(['[%s] ' fmt '\n'], datestr(now,'HH:MM:SS'), varargin{:});
end

File: calibration/ctl/run_generation.py
#!/usr/bin/env python3
"""常驻标定驱动 —— run_generation.m 的编译版对应物。

和 .m 版的区别只有一处: 撒腿的方式。
    .m 版:  一个 MATLAB 持 parpool(120), parfeval 撒腿   -> 要 1 MATLAB + 1 PCT 席位
    本版:   fork 120 个 MCR standalone 进程, 各跑一条腿   -> 0 席位

其余语义逐条照搬, 不做"顺便改进":
  每代 = 读 state.json 的 batchNN_cases -> 撒 nCase*ndose 条腿 -> 全部回收
         -> 逐 case 抽指标 -> 判成功 -> CMA 出下一代 -> 继续
  已完成的腿跳过 (续算); 反复失败的 case 熔断; 墙钟耗尽干净退出等续投。

CMA/目标函数仍然全在 python (gen_tool.py / propose_cmaes.py), 编译不影响它们 ——
改目标函数不用重编。

环境变量 (与 .m 版同名同义):
  MATFDM_RUN         运行目录 (数据根)
  CALIB_PYTHON       python3
  CALIB_POPULATION / CALIB_WORKERS / CALIB_ENDPOINT_TOL / CALIB_MAX_ATTEMPTS
  CALIB_WALL_BUDGET  本作业可用秒数 (不设 = 不自我限时, 跑到 SLURM 杀)
  CALIB_GEN_RESERVE  开新一代所需的最少剩余秒数 (默认 5400)
  MATFDM_DEADLINE    作业截止 epoch 秒 (multi_node.sh 从 scontrol 算好传入)
"""

from __future__ import annotations

import json
import os
import subprocess
import sys
import time
from datetime import datetime
from pathlib import Path

HERE = Path(__file__).resolve().parent
MCR = HERE / "matfdm_mcr.sh"
GEN_TOOL = HERE / "gen_tool.py"


def say(fmt, *a):
    print(f"[{datetime.now():%H:%M:%S}] " + (fmt % a if a else fmt), flush=True)


def envnum(name, default):
    try:
        return float(os.environ[name])
    except (KeyError, ValueError):
        return default


# ------------------------------------------------------------------ 状态
def read_cases(statef: Path, batch: int):
    s = json.loads(statef.read_text())
    key = f"batch{batch:02d}_cases"
    if key not in s:
        raise RuntimeError(f"state.json 缺 {key}")
    return s[key]


def next_batch(statef: Path, metrics: Path) -> int:
    """已有的最大 batch: 若其 cases 已全部抽完指标则 +1, 否则续跑它。"""
    if not statef.is_file():
        return 1
    s = json.loads(statef.read_text())
    nums = [int(k[5:7]) for k in s
            if k.startswith("batch") and k.endswith("_cases")]
    if not nums:
        return 1
    b = max(nums)
    for c in read_cases(statef, b):
        if not (metrics / f"{c['case_tag']}.csv").is_file():
            return b            # 本代还没做完, 续跑本代
    return b + 1


def leg_is_complete(run: Path, tag: str, dose) -> bool:
    """与 leg_is_complete.m 同一判据: _COMPLETE 标记 + 四个非空 final csv。"""
    root = run / "decouple" / tag / f"dose{dose:g}"
    if not (root / "_COMPLETE").is_file():
        return False
    for name in ("Cr2O3", "Fe3O4", "FeCr2O4", "SiO2"):
        f = root / f"{name}_final.csv"
        if not f.is_file() or f.stat().st_size == 0:
            return False
    return True


# ------------------------------------------------------------------ 外部调用
def gen_tool(run: Path, cfg: dict, *args) -> tuple[int, str]:
    """调 gen_tool.py; 环境与 .m 版的 sh() 一致。

    CALIB_POPULATION 必须显式传 —— propose_cmaes 拿它和磁盘上的 case 数对账,
    不传就用默认值, 一代 40 个 case 会被判成"与 POPULATION 不符"直接拒绝。
    判据类的几个 (MIDDLE_MAX/ENDPOINT_BAND/SEED) 正常由 node_task.sh 导出;
    这里在缺失时用 config.json 兜底, 好让脚本单独跑也对。
    """
    env = dict(os.environ)
    env.update(MATFDM_RUN=str(run), CALIB_REPO_DIR=str(run),
               CALIB_STATE_PATH=str(run / "calibration" / "state.json"),
               CALIB_OUT_DIR=str(run / "calibration" / "optimizer"),
               CALIB_POPULATION=str(int(envnum("CALIB_POPULATION",
                                               cfg.get("population", 40)))))
    for var, key, default in (("CALIB_MIDDLE_MAX", "middle_max", 70),
                              ("CALIB_ENDPOINT_BAND", "endpoint_band", 5),
                              ("CALIB_SEED", "seed", 20260804)):
        env.setdefault(var, str(cfg.get(key, default)))
    py = os.environ.get("CALIB_PYTHON", "python3")
    p = subprocess.run([py, str(GEN_TOOL), *map(str, args)],
                       env=env, capture_output=True, text=True)
    return p.returncode, p.stdout + p.stderr


def spawn_leg(run: Path, tag: str, dose, mult, budget) -> subprocess.Popen:
    m = ",".join(repr(float(x)) for x in mult)
    cmd = [str(MCR), "leg", str(run), tag, f"{dose:g}", m]
    if budget is not None:
        cmd.append(str(int(budget)))
    log = run / "calibration" / "logs" / f"{tag}_dose{dose:g}.out"
    return subprocess.Popen(cmd, stdout=log.open("w"), stderr=subprocess.STDOUT)


def run_metrics(run: Path, tag: str) -> bool:
    p = subprocess.run([str(MCR), "metrics", str(run), tag],
                       capture_output=True, text=True)
    return p.returncode == 0


# ------------------------------------------------------------------ 熔断计数
def load_attempts(f: Path) -> dict:
    try:
        return json.loads(f.read_text())
    except (OSError, ValueError):
        return {}


def save_attempts(f: Path, a: dict):
    tmp = f.with_suffix(".json.tmp")
    tmp.write_text(json.dumps(a))
    tmp.replace(f)


# ------------------------------------------------------------------ 主循环
def main() -> int:
    run = Path(os.environ.get("MATFDM_RUN", ""))
    if not run.is_dir():
        say("MATFDM_RUN 没设或不存在: %s", run)
        return 2
    calib = run / "calibration"
    statef = calib / "state.json"
    metrics = calib / "metrics"
    for d in (metrics, calib / "logs", calib / "optimizer"):
        d.mkdir(parents=True, exist_ok=True)

    cfg = json.loads((run / "config.json").read_text())
    pop = int(envnum("CALIB_POPULATION", cfg.get("population", 40)))
    nleg = int(envnum("CALIB_WORKERS", cfg.get("workers", 120)))
    tol = envnum("CALIB_ENDPOINT_TOL", cfg.get("endpoint_tol", 3))
    max_attempts = int(envnum("CALIB_MAX_ATTEMPTS", cfg.get("max_attempts", 3)))
    doses = list(cfg.get("doses", [0, 0.5, 3]))

    budget = envnum("CALIB_WALL_BUDGET", float("inf"))
    reserve = envnum("CALIB_GEN_RESERVE", 5400)
    deadline = envnum("MATFDM_DEADLINE", 0)
    t0 = time.time()

    say("[boot] run=%s legs<=%d pop=%d %s", run, nleg, pop,
        f"budget={budget:.0f}s" if budget != float("inf") else "无代码侧时限")
    say("[boot] 引擎=MCR standalone (0 个 MATLAB/PCT 席位), build=%s",
        os.environ.get("MATFDM_BUILD_COMMIT", "?"))

    attempts_f = calib / "attempts.json"

    while True:
        if (calib / "DONE").is_file() or (calib / "STOP").is_file():
            say("[sentinel] 见到 DONE/STOP, 退出")
            break
        if budget != float("inf") and time.time() - t0 > budget - reserve:
            say("[wall] 剩余不足一代, 干净退出")
            break

        batch = next_batch(statef, metrics)

        rc, hist = gen_tool(run, cfg, "history", batch)
        print(hist, end="", flush=True)

        rc, out = gen_tool(run, cfg, "propose", batch)
        if rc != 0:
            say("[propose] rc=%d, python 输出:", rc)
            print(out, flush=True)
            if batch > 1:
                say("[propose] 失败, 惩罚上一代缺指标的 case 后重试")
                gen_tool(run, cfg, "penalize-missing", batch - 1)
                rc, out = gen_tool(run, cfg, "propose", batch)
        if rc != 0:
            say("[propose] CMA-ES 提议失败, batch %d", batch)
            return 3

        cases = read_cases(statef, batch)
        say("[gen %02d] %d cases x %d dose = %d legs",
            batch, len(cases), len(doses), len(doses) * len(cases))

        # ---- 撒腿 (已完成的跳过, 这就是续算) ----
        if budget != float("inf"):
            leg_budget = max(600, budget - (time.time() - t0) - 300)
        elif deadline > 0:
            leg_budget = max(600, deadline - time.time() - 300)
        else:
            leg_budget = None

        pending = [(c["case_tag"], d, c["mult"])
                   for c in cases for d in doses
                   if not leg_is_complete(run, c["case_tag"], d)]
        say("[gen %02d] 待跑 %d 条 (已完成 %d 条直接跳过)",
            batch, len(pending), len(doses) * len(cases) - len(pending))

        procs, meta, queue = {}, {}, list(pending)
        done_n, t_gen = 0, time.time()
        attempts = load_attempts(attempts_f)

        while queue or procs:
            while queue and len(procs) < nleg:
                tag, d, mult = queue.pop(0)
                p = spawn_leg(run, tag, d, mult, leg_budget)
                procs[p.pid] = p
                meta[p.pid] = (tag, d)
            if not procs:
                break
            time.sleep(5)
            for pid in [q for q, p in procs.items() if p.poll() is not None]:
                p = procs.pop(pid)
                tag, d = meta.pop(pid)
                done_n += 1
                if p.returncode == 0:
                    st = "complete"
                elif p.returncode == 10:
                    st = "wallclock"
                else:
                    st = f"error(rc={p.returncode})"
                    k = f"{tag}/dose{d:g}".replace("/", "_").replace(".", "_")
                    attempts[k] = attempts.get(k, 0) + 1
                    save_attempts(attempts_f, attempts)
                say("[leg] %-24s %-10s (%d/%d, 本代 %.0f min)",
                    f"{tag}/dose{d:g}", st, done_n, len(pending),
                    (time.time() - t_gen) / 60)

        # ---- 熔断 ----
        dead = []
        for c in cases:
            tag = c["case_tag"]
            for d in doses:
                k = f"{tag}/dose{d:g}".replace("/", "_").replace(".", "_")
                if not leg_is_complete(run, tag, d) and attempts.get(k, 0) >= max_attempts:
                    dead.append(tag)
                    break
        for tag in dict.fromkeys(dead):
            gen_tool(run, cfg, "penalize", batch, "--tag", tag)
            say("[dead] %s 连续失败 >=%d 次, 已写惩罚指标", tag, max_attempts)

        # ---- 完整性: 有腿没跑完且未熔断 = 墙钟耗尽, 交给续投 ----
        incomplete = sum(1 for c in cases if c["case_tag"] not in dead
                         for d in doses if not leg_is_complete(run, c["case_tag"], d))
        if incomplete > 0:
            say("[gen %02d] %d 条腿未完成 (墙钟), 存盘退出等续投", batch, incomplete)
            break

        # ---- 抽指标 ----
        for c in cases:
            tag = c["case_tag"]
            if (metrics / f"{tag}.csv").is_file():
                continue
            if not run_metrics(run, tag):
                say("[metrics] %s 失败, 写惩罚指标", tag)
                gen_tool(run, cfg, "penalize", batch, "--tag", tag)

        # ---- 判成功 ----
        rc, out = gen_tool(run, cfg, "success", batch, "--tol", tol)
        say("[gen %02d] %s", batch, out.strip())
        winners = ""
        for line in out.splitlines():
            if line.startswith("WINNERS="):
                winners = line[len("WINNERS="):].strip()
        if rc == 0 and winners:
            (calib / "DONE").write_text(winners + "\n")
            say("[done] 命中: %s", winners)
            break
        rc, rep = gen_tool(run, cfg, "report", batch)
        print(rep, end="", flush=True)

    say("[exit] 总耗时 %.2f h", (time.time() - t0) / 3600)
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
